### DISCUSSION TITLE
mortar interface for accounting for stress discontinuity around a fault.

### DISCUSSION THREAD

**Discussion by Thomas-Ulrich** [idaholab/moose#26574] [Q&A General] [answered]
Hi everyone,

We are working on coupling THM models of geothermal systems with earthquake models.
In the THM application (application Tiger https://github.com/Domino2022/Tiger, based on the moose framework), we are modeling fluid injection within a 3D reservoir including faults, which are lower-dimensional features defined in the mesh (e.g. allowing to model the enhance fluid diffusivity at the fault). The mesh is built with gmsh.

When the conditions for an earthquake are reached, we pass the stress state to the external earthquake modeling software (a discontinuous Galerkin code, https://github.com/SeisSol/SeisSol) and model the earthquake. Fault slip during the earthquake leads to a displacement discontinuity across the fault and comes in pairs with a stress change within the volume, also discontinuous across the fault.

Currently, when porting back the stress change to the THM model through a SolutionUserObject, we get a smooth version of the discontinuity. Specifically, the tractions on the fault are wrong, because based on interpolated stress instead of calculated from the stress on the 2 sides.
 
 To solve this problem, I would duplicate the fault nodes and triangles, to create "split nodes", and then glue them with a mortar element in the THM model. Do you think this would work?
 Is there an automated way to generate a duplicated internal surface with gmsh for mortar application?
 Would a EqualValueConstraint be sufficient to capture the discontinuity of the stress at the fault interface when retstarting moose?

**Comment by lindsayad** [comment]
You could create sidesets for each side of the interface, and then you could use a `LowerDBlockFromSidesetGenerator` to create the lower-dimensional blocks associated with the sidesets which mortar needs. What variable(s) would you be appying the `EqualValueConstraint` to?

**Reply by GiudGiud** [reply, marked as answer]
To break the two sides apart, you can also use this mesh generator
https://mooseframework.inl.gov/source/meshgenerators/BreakMeshByBlockGenerator.html

if the sideset is in between two blocks. If it is not, then I would recommend first trying to create the two blocks on both sides of the sideset.
If that is impossible, then I would look at creating a new object: BreakMeshOnSidesetGenerator

**Reply by Thomas-Ulrich** [reply]
Thank you! That's very useful.
Well I think I will first use EqualValueConstraint on all variables, and see how much it differs from the current setup.

**Reply by lindsayad** [reply]
This input demonstrates that mortar can be used even when the higher-dimensional face is shared by the secondary and primary sides, e.g. without breaking the mesh connectivity
```
[Mesh]
  [file]
    type = GeneratedMeshGenerator
    xmax = 2
    nx = 20
    ny = 10
    dim = 2
  []
  [right_block]
    input = file
    type = SubdomainBoundingBoxGenerator
    bottom_left = '1 0 0'
    top_right = '2 1 0'
    block_id = '1'
    block_name = 'right'
  []
  [left_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = '0'
    paired_block = 'right'
    input = right_block
    new_boundary = 'interface_left'
  []
  [right_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = 'right'
    paired_block = '0'
    input = left_sideset
    new_boundary = 'interface_right'
  []
  [left_lowerd_block]
    input = right_sideset
    type = LowerDBlockFromSidesetGenerator
    sidesets = 'interface_left'
    new_block_name = 'left_lower'
  []
  [right_lowerd_block]
    input = left_lowerd_block
    type = LowerDBlockFromSidesetGenerator
    sidesets = 'interface_right'
    new_block_name = 'right_lower'
  []
  second_order = true
[]

[Variables]
  [u]
    order = SECOND
    family = LAGRANGE
    block = '0'
  []
  [v]
    order = SECOND
    family = LAGRANGE
    block = 'right'
  []

  [lm]
    order = FIRST
    family = LAGRANGE
    block = 'left_lower'
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
  [diff_v]
    type = Diffusion
    variable = v
  []
[]

[Constraints]
  [ced]
    type = EqualValueConstraint
    variable = lm
    secondary_variable = u
    primary_variable = v
    primary_boundary = interface_right
    primary_subdomain = right_lower
    secondary_boundary = interface_left
    secondary_subdomain = left_lower
  []
[]

[BCs]
  [left]
    type = DirichletBC
    variable = u
    boundary = 'left'
    value = 0
  []
  [right]
    type = DirichletBC
    variable = v
    boundary = 'right'
    value = 1
  []
[]


[Executioner]
  solve_type = NEWTON
  type = Steady
  nl_rel_tol = 1e-11
  petsc_options_iname = '-pc_type -pc_factor_shift_type'
  petsc_options_value = 'lu       NONZERO'
[]

[Outputs]
  exodus = true
[]
```

**Comment by Rahim-Habibi** [comment]
> This input demonstrates that mortar can be used even when the higher-dimensional face is shared by the secondary and primary sides, e.g. without breaking the mesh connectivity
> 
> ```
> [Mesh]
>   [file]
>     type = GeneratedMeshGenerator
>     xmax = 2
>     nx = 20
>     ny = 10
>     dim = 2
>   []
>   [right_block]
>     input = file
>     type = SubdomainBoundingBoxGenerator
>     bottom_left = '1 0 0'
>     top_right = '2 1 0'
>     block_id = '1'
>     block_name = 'right'
>   []
>   [left_sideset]
>     type = SideSetsBetweenSubdomainsGenerator
>     primary_block = '0'
>     paired_block = 'right'
>     input = right_block
>     new_boundary = 'interface_left'
>   []
>   [right_sideset]
>     type = SideSetsBetweenSubdomainsGenerator
>     primary_block = 'right'
>     paired_block = '0'
>     input = left_sideset
>     new_boundary = 'interface_right'
>   []
>   [left_lowerd_block]
>     input = right_sideset
>     type = LowerDBlockFromSidesetGenerator
>     sidesets = 'interface_left'
>     new_block_name = 'left_lower'
>   []
>   [right_lowerd_block]
>     input = left_lowerd_block
>     type = LowerDBlockFromSidesetGenerator
>     sidesets = 'interface_right'
>     new_block_name = 'right_lower'
>   []
>   second_order = true
> []
> 
> [Variables]
>   [u]
>     order = SECOND
>     family = LAGRANGE
>     block = '0'
>   []
>   [v]
>     order = SECOND
>     family = LAGRANGE
>     block = 'right'
>   []
> 
>   [lm]
>     order = FIRST
>     family = LAGRANGE
>     block = 'left_lower'
>   []
> []
> 
> [Kernels]
>   [diff]
>     type = Diffusion
>     variable = u
>   []
>   [diff_v]
>     type = Diffusion
>     variable = v
>   []
> []
> 
> [Constraints]
>   [ced]
>     type = EqualValueConstraint
>     variable = lm
>     secondary_variable = u
>     primary_variable = v
>     primary_boundary = interface_right
>     primary_subdomain = right_lower
>     secondary_boundary = interface_left
>     secondary_subdomain = left_lower
>   []
> []
> 
> [BCs]
>   [left]
>     type = DirichletBC
>     variable = u
>     boundary = 'left'
>     value = 0
>   []
>   [right]
>     type = DirichletBC
>     variable = v
>     boundary = 'right'
>     value = 1
>   []
> []
> 
> 
> [Executioner]
>   solve_type = NEWTON
>   type = Steady
>   nl_rel_tol = 1e-11
>   petsc_options_iname = '-pc_type -pc_factor_shift_type'
>   petsc_options_value = 'lu       NONZERO'
> []
> 
> [Outputs]
>   exodus = true
> []
> ```
Dear @lindsayad, 
in this example, you applied the constraint on main variables. 
I have looked at the example [here](https://mooseframework.inl.gov/source/constraints/EqualValueConstraint.html), all of them operate on main variable. Also, according to . C file, it operates on "_u_primary and _u_secondary".  
So, I was wondering to know is it possible to apply it on axuvariable, e.g. stress?

**Reply by GiudGiud** [reply]
Stress is tensor mechanics is actually a material property. 
Are you trying to constrain stresses to match on both sides of a boundary? Why not strain / displacements?

**Comment by Rahim-Habibi** [comment]
Thanks @GiudGiud 
Actually we have a surface with 2D element mesh (fault in this case) between two subdomain . So, we want to force the stress continiuty from sidessets of the subdomains to the surface. 
I have already tried with displacement, it works. But based on this issue #24897, lower dimensional block should be removed from "module" and "kernel" blocks (i.e. it has to be called only in "Constraint" block. So. my understanding is that the stress belong to lower dimensional can not be called in module and kernel meaning that it only forces displacement continuity, right?
that is why I asked is there a way to have stress continuty on the lower dimensional block using constraint. 
Thanks 
Rahim

**Reply by GiudGiud** [reply]
> stress continuty on the lower dimensional block

what do you mean by that? continuous from the volumes (both sides) onto the lower dimensional block?
Or continuous across the sideset (between the two volumes)

**Comment by Rahim-Habibi** [comment]
> > stress continuty on the lower dimensional block
> 
> what do you mean by that? continuous from the volumes (both sides) onto the lower dimensional block? Or continuous across the sideset (between the two volumes)

@GiudGiud 
According to #24897, it is mentioned that lower dimensional block has to be removeed from "tensor mechanics module" and "kernel" blocks (meaning that stress is not going to be compted on lower dimensional block).
I followed the example there and successfully i got the displacement contitiuty as expected but stress tended to zero on lower dimensional block since based on the example, stress on lower dimensional block is not called in "tensor mechanics module" and "kernel" blocks (so it makes sense not to have stress on lower dimensional block). But I want to have stress continuty in my simulation that is why I am wondering to know how to have stress continuty on a lower dimensional block surrounded by two subdomains.

So, here my mesh and constraints blocks to have displacement continuty which works well:
```
[Mesh]
  [./file]
    type = FileMeshGenerator
    file = tpv104-inbox-fault.msh
  [../]
  [./sidesetmid1]
    type = SideSetsBetweenSubdomainsGenerator
    input = 'file'
    primary_block = 'SDMidBack'                  # subdomain
    paired_block = 'SDMidFront'                   # subdomain
    new_boundary = 'primary1_interface'     # sideset
  [../]
  [./sidesetmid2]
    type = SideSetsBetweenSubdomainsGenerator
    input = 'sidesetmid1'
    primary_block = 'SDMidFront'                # subdomain
    paired_block = 'SDMidBack'                   # subdomain
    new_boundary = 'primary2_interface'    # sideset
  [../]
  [interfacemid]
    type = LowerDBlockFromSidesetGenerator
    input = sidesetmid2
    sidesets = 'primary2_interface'               # sideset
    new_block_id = '10000'
    new_block_name = 'interfacemid'          # lower dimensional block
    []
    []
##########
[Modules]
  [./TensorMechanics]
    [./Master]
      [./all]
      add_variables = true
      strain = SMALL
      incremental = true
      #temperature = temperature
      #use_displaced_mesh = true
      eigenstrain_names = 'reduced_eigenstrain'
      volumetric_locking_correction = true
      additional_generate_output = 'stress_xx stress_xy stress_xz stress_yx stress_yy stress_yz stress_zx stress_zy stress_zz'
      block = 'box SDRightBack SDRightFront SDLeftBack SDLeftFront SDBottomBack SDBottomFront SDMidBack SDMidFront '  # lower dimensional should be removed from here if you want use Constraint on lower dimensional block
      [../]
    [../]
  [../]
 []
###############
[Constraints]
  [./mortarz]
    type = EqualValueConstraint  
    primary_boundary = SDMidBack
    secondary_boundary = SDMidFront
    primary_subdomain = primary1_interface
    secondary_subdomain = primary2_interface
    secondary_variable = disp_z
    primary_variable = disp_z
    variable = disp_z
  [../]
[]

BUT, when I change the constraint to work on "stress_zz" for example, 
[Constraints]
  [./mortarz]
    type = EqualValueConstraint  
    primary_boundary = SDMidBack
    secondary_boundary = SDMidFront
    primary_subdomain = primary1_interface
    secondary_subdomain = primary2_interface
    secondary_variable = stress_zz
    primary_variable = stress_zz
    variable = stress_zz
  [../]
[]
```
I got the following error, at two scenarios with/without calling stress_zz in module block:

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

No nonlinear variable named stress_zz found. Did you specify an auxiliary variable when you meant to specify a nonlinear variable?
```

**Reply by GiudGiud** [reply]
Hello

stress_zz is a material property, that's why it wont let you specify it in a parameter that expects a variable in the `EqualValueConstraint`. 
I dont see anything obvious that does what you want in the existing constraints:
https://mooseframework.inl.gov/syntax/Constraints/index.html
but maybe take another look

I think at this point you want to write your own constraint that sets a material property to be continuous, more precisely a component of a tensor material property.
I would use automatic differentiation. You write the contribution to the residual just like any other constraint, and AD will do the Jacobian. Jacobian would be tricky otherwise because you'd need to write the derivatives of the stress with regards to the displacement (your nonlinear variable).

Guillaume

**Reply by lindsayad** [reply]
Continuity of stress is conceptually similar to continuity of heat flux which is implemented in the mortar constraints in the `heat_transfer` module. The nice thing about these conditions is that they do not introduce saddle points like `EqualValueConstraint`

**Comment by Rahim-Habibi** [comment]
Thanks for your reply @GiudGiud  and @lindsayad 


> Hello
> 
> stress_zz is a material property, that's why it wont let you specify it in a parameter that expects a variable in the `EqualValueConstraint`. I dont see anything obvious that does what you want in the existing constraints: https://mooseframework.inl.gov/syntax/Constraints/index.html but maybe take another look
> 
> I think at this point you want to write your own constraint that sets a material property to be continuous, more precisely a component of a tensor material property. I would use automatic differentiation. You write the contribution to the residual just like any other constraint, and AD will do the Jacobian. Jacobian would be tricky otherwise because you'd need to write the derivatives of the stress with regards to the displacement (your nonlinear variable).
> 
> Guillaume

Thanks for your reply @GiudGiud  and @lindsayad 
I have changed the method, instead of using constraints I decided to use interface kernels. From "InterfaceKernels", I have tried InterfaceDiffusion and got reasonable results. I have disp and stress continuty on the lower dimensioanl block. 
it seems it is easy to work with w.r.t constraints.

**Reply by lindsayad** [reply]
yes if you have a conforming mesh at your interface, interface kernels are absolutely the way to go
